Added support for pulsatrix chargers

Sauttets opened the pull request.

Sauttets committed:
Create pulsatrix.go

Sauttets committed:
Update brands.json

Sauttets committed:
Create pulsatrix.yaml

Sauttets committed:
Update definitions.go

Sauttets committed:
Update definitions.go

added new line at eof

Sauttets committed:
Merge pull request #1 from Sauttets/pulsatrix-Charger

Added a Pulsatrix charger

Sauttets committed:
added new line at eof

Sauttets committed:
added new line at eof

Sauttets committed:
removed unused variable

Sauttets committed:
removed path

andig reviewed (commented):
Great PR- I think this can still be simplified quite a bit. Hope you finde the comments valuable.

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,289 @@
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/gorilla/websocket"
```
We're using

    github.com/nhooyr/websocket

Sauttets replied:
i was using gorilla since both WS-libs are in the go.mod file, but will change

Sauttets commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,289 @@
+}
+
+// wsWrite writes a message to the websocket
+func (c *PulsatrixCharger) wsWrite(message []byte) error {
```
The send function wants []byte anyways. Is there any benefit of converting it multiple times or am I missing something?

andig replied:
You can then omit the byte conversion at all call sites. Doesn't matter otherwise.

Sauttets commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,289 @@
+func (c *PulsatrixCharger) parseWsMessage(messageType int, message []byte) {
+	if messageType == websocket.TextMessage {
+		b := bytes.Replace(message, []byte(":NaN"), []byte(":null"), -1)
+		stadtIdx := bytes.IndexByte(b, '{')
```
the issue is that there are 3 different messages with different headers. 

```
configuration:
{"session-id":"018b94bc-5301-51b0-bd12-bce56db6317f","effectiveAmperageLimit":32}
charge: 
{"voltage":[234.9605713,234.8770905,235.0018616],"amperage":[0,0,0],"frequency":49.94384384,"energyImported":0.302999973,"allocatedAmperage":0,"commandedAmperage":16,"lastActivePower":0,"lastMeterValue":0.302999973,"availableAmperage":0,"signaledAmperage":0}
transaction:
{"state":"IDLE","startedTime":1699007451,"endedTime":1699007614,"meterStart":NaN,"meterStop":0.302999973,"vehicleStatus":"A","chargeControllerStatus":"A1","usedPhasesSession":"000","startReason":"CablePluggedIn","stopReason":"StoppedByEV","allocationSourceName":"Transaction"}
```

andig replied:
Is the `configuration:` part of the message? Thats ugly!
If not, it's json already and no need to look for start of message?

Sauttets replied:
yes, it's part of the message but since i am not able to change this, i have to work with it

Sauttets commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,289 @@
+func (c *PulsatrixCharger) heartbeat() {
+	for {
+		if c.VehicleStatus == "A" {
+			c.handleError(c.Enable(false))
```
if the charging controller doesn't get a request at least once every 300s it will mark external data invalid and take back the control

Sauttets commented:
Thanks for the amazing comments, this really helps

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
improved performance and security

Sauttets committed:
Removed unnecessary description

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+
+// ConnectWs connects to a pulsatrix SECC websocket
+func (c *PulsatrixCharger) connectWs(hostname string) error {
+	uri := fmt.Sprintf("ws://%s/api/ws", hostname)
```
I would also make uri a charger property, not host?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+func (c *PulsatrixCharger) connectWs(hostname string) error {
+	uri := fmt.Sprintf("ws://%s/api/ws", hostname)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
```
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), request.Timeout)
```

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+
+	if err != nil {
+		if strErr := fmt.Sprintf("%s", err); strErr == "websocket: bad handshake" {
+			c.log.ERROR.Println("bad handshake. Make sure the IP and port are correct")
```
Handle errors only once. Either log or return, not both.

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+
+// Heartbeat sends a heartbeat to the pulsatrix SECC
+func (c *PulsatrixCharger) heartbeat() {
+	for {
```
```suggestion
	for range time.Tick(3 * time.Minute) {
```

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+		} else {
+			c.handleError(c.Enable(false))
+		}
+		time.Sleep(3 * time.Minute)
```
```suggestion
```

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+
+// Status implements the api.Charger interface
+func (c *PulsatrixCharger) Status() (api.ChargeStatus, error) {
+	return api.ChargeStatusString(c.VehicleStatus)
```
Mutex!

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+
+// GetMaxCurrent implements the api.CurrentGetter interface
+func (c *PulsatrixCharger) GetMaxCurrent() (float64, error) {
+	return float64(c.AllocatedAmperage), nil
```
Mutex!

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+
+// CurrentPower implements the api.Meter interface
+func (c *PulsatrixCharger) CurrentPower() (float64, error) {
+	return float64(c.LastActivePower), nil
```
Mutex!

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+// StartCharge implements the api.VehicleChargeController interface
+func (c *PulsatrixCharger) StartCharge() error {
```
Why? Does that force any charging to start beyond what `Enable` already does?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+// StopCharge implements the api.VehicleChargeController interface
+func (c *PulsatrixCharger) StopCharge() error {
```
Dito?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+
+// Total Energy implements the api.MeterEnergy interface
+func (c *PulsatrixCharger) TotalEnergy() (float64, error) {
+	return float64(c.EnergyImported), nil
```
Mutex!

andig commented on cmd/detect/definitions.go:
```diff
@@ -39,6 +39,7 @@ const (
 	taskFroniusWeb   = "fronius-web"
 	taskTasmota      = "tasmota"
 	taskShelly       = "shelly"
+	taskPulsatrix    = "pulsatrix"
```
Does that detect anything? I don't think we should add an "open port 80" detection here.

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+
+// WsReader runs a loop that reads messages from the websocket
+func (c *PulsatrixCharger) wsReader() {
+	var breaker bool
```
Again: why?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+					c.log.ERROR.Println("error reading message: ", err)
+					c.tryRead = 0
+					c.reconnectWs()
+					breaker = true
```
return

would be so much simpler?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+		for !breaker {
+			messageType, message, err := c.conn.Read(context.Background())
+			if err != nil {
+				if c.tryRead < 3 {
```
why not use Backoff?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+// ReconnectWs reconnects to a pulsatrix SECC websocket
+func (c *PulsatrixCharger) reconnectWs() {
```
The whole reconnect logic still feels much to complex for stable operation. It's not obvious to me if this is really synchronized or racy?

Sauttets committed:
Merge branch 'evcc-io:master' into pulsatrix-Charger

Sauttets committed:
removed unnessecary panic recovery

Sauttets committed:
improved charger logic

Sauttets committed:
removed unnessecary open port 80

Sauttets committed:
fixed writeWs logic

Sauttets committed:
fixed RC in reconnect logic

Sauttets committed:
Merge branch 'evcc-io:master' into pulsatrix-Charger

Sauttets committed:
removed unnecessary description

Sauttets committed:
lint and timeout

Sauttets committed:
Merge branch 'evcc-io:master' into master

Sauttets committed:
Update pulsatrix.go

Sauttets committed:
Update definitions.go

Sauttets commented:
I am sorry if i make this unnecessarily complex, i am still at the start of my career, i hope you can understand :)

Sauttets committed:
removed duplicate error handling

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
newline eof confusion

Sauttets committed:
Update brands.json

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+			ctx, cancel := context.WithTimeout(context.Background(), request.Timeout)
+			defer cancel()
+			messageType, message, err := c.conn.Read(ctx)
+			if err != nil {
```
if errors.Is(err, context.DeadlineExceeded) {

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+	c.log.INFO.Printf("connecting to %s", c.uri)
+	conn, _, err := websocket.Dial(ctx, c.uri, nil)
+	if err != nil {
+		if strErr := fmt.Sprintf("%s", err); strings.Contains(strErr, "context deadline exceeded") {
```
if errors.Is(err, context.DeadlineExceeded) {

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+
+// WsReader runs a loop that reads messages from the websocket
+func (c *PulsatrixCharger) wsReader() {
+	go func() {
```
Why the double go func?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+func (c *PulsatrixCharger) parseWsMessage(messageType websocket.MessageType, message []byte) {
+	if messageType == websocket.MessageText {
+		b := bytes.ReplaceAll(message, []byte(":NaN"), []byte(":null"))
+		stadtIdx := bytes.IndexByte(b, '{')
```
What is `stadtIdx`?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+// Heartbeat sends a heartbeat to the pulsatrix SECC
+func (c *PulsatrixCharger) heartbeat() {
+	for range time.Tick(3 * time.Minute) {
+		if c.enState {
```
c.handleError(c.Enable(c.enState))

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+}
+
+func (c *PulsatrixCharger) valid() bool {
+	return !c.updated.Before(time.Now().Add(-30 * time.Second))
```
```suggestion
	return time.Since(c.updated) < 30 * time.Second
```

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+	return !c.updated.Before(time.Now().Add(-30 * time.Second))
+}
+
+func (c *PulsatrixCharger) handleError(err error) {
```
do we need this as function?

Sauttets replied:
I thought this would allow for a cleaner code

andig replied:
most case you log and return error. logging _is_ error handling, same as returning. you want to handle it only once. If you remove the logging there won't be much uses for this function?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+
+func (c *PulsatrixCharger) handleError(err error) {
+	if err != nil {
+		c.log.ERROR.Println("error:", err)
```
```suggestion
		c.log.ERROR.Println(err)
```

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+		c.handleError(c.Enable(true))
+	}
+	res := strconv.FormatFloat(current, 'f', 10, 64)
+	return c.wsWriter("setCurrentLimit\n" + res)
```
```suggestion
	return c.wsWriter("setCurrentLimit\n" + strconv.FormatFloat(current, 'f', 10, 64))
```

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if !c.valid() {
+		c.handleError(c.Enable(false))
```
Either log or return. Not both. Use

    api.ErrOutdated

Sauttets committed:
Update charger/pulsatrix.go

Co-authored-by: andig <cpuidle@gmx.de>

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+	defer c.mutex.Unlock()
+	currents := c.PhaseAmperage
+	if len(currents) < 3 || !c.valid() {
+		c.handleError(c.Enable(false))
```
log or return

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+	defer c.mutex.Unlock()
+	voltages := c.PhaseVoltage
+	if len(voltages) < 3 || !c.valid() {
+		c.handleError(c.Enable(false))
```
dito...

Sauttets committed:
better error handling charger/pulsatrix.go

Co-authored-by: andig <cpuidle@gmx.de>

Sauttets committed:
Update charger/pulsatrix.go

Co-authored-by: andig <cpuidle@gmx.de>

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+		}
+		return err
+	}
+	c.updated = time.Now()
```
I don't think it has been updated at this time?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+	c.updated = time.Now()
+	c.enState = false
+	c.conn = conn
+	c.handleError(c.Enable(false))
```
this would disable right away- is that really what you want?

Sauttets replied:
i added this to ensure the evcc and the charging controller are in sync even after a error occurred

andig replied:
Still, if this happens you will hard-disable the charger every time. If currently charging, that is not a good idea. Let the loadpoint control the state. Charger implementations should not introduce additional logic.

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,269 @@
+
+// ReconnectWs reconnects to a pulsatrix SECC websocket
+func (c *PulsatrixCharger) reconnectWs() {
+	c.handleError(backoff.RetryNotify(func() error {
```
```suggestion
	c.handleError(backoff.RetryNotify(c.connectWs),
```

Sauttets replied:
i am not sure what you mean with that. Is this what you wanted?
```suggestion
		notify := func(err error, time time.Duration) {
		c.log.WARN.Printf("trying to reconnect in %v...\n", time)
	}
	c.handleError(backoff.RetryNotify(c.connectWs, c.bo, notify)) 
	return
```

andig replied:
You don't need the wrapper around `c.connectWs`, just remove it.

Sauttets committed:
small fixes

Sauttets committed:
small logic changes

Sauttets committed:
Merge branch 'master' into pulsatrix-Charger

Sauttets committed:
Merge pull request #3 from Sauttets/pulsatrix-Charger

Pulsatrix charger

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,272 @@
+func (c *PulsatrixCharger) MaxCurrentMillis(current float64) error {
+	c.signaledAmp = current
+	if !c.enState && current > 0 {
+		c.handleError(c.Enable(true))
```
Errh. Why should reading the current enable the charger? Don't do this!

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,272 @@
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if !c.valid() {
+		c.handleError(c.Enable(false))
```
dito

andig replied:
And again: either log or return the error. not both.

Sauttets replied:
as I said above, those are 2 different errors:

```suggestion
		c.handleError(c.Enable(false))
		return 0, fmt.Errorf(api.ErrOutdated.Error())
```

the fist line (the log) handles the error if the c.Enable function call fails
the second one (the return) is for the outdated data and a different error

andig commented:
I can offer to do a full cleanup. However, I cannot push to your branch.

Sauttets committed:
Merge branch 'evcc-io:master' into master

Sauttets committed:
Merge branch 'evcc-io:master' into pulsatrix-Charger

Sauttets commented:
@andig 
Da du ja auch Deutscher bist, der Einfachheit halber auf Deutsch:
Es wäre sehr nett, wenn du ein Cleanup machen könntest. Du kannst allerdings aktuell nicht auf meinen Branch pushen, weil ich dir die Rechte nur bei einem offenen PR geben kann.
Ich habe mein Bestes gegeben, um deine Vorschläge zu implementieren, scheine damit allerdings nicht wirklich viel Erfolg gehabt zu haben. Was wäre denn jetzt dein Vorschlag, um den Ladepunkt in deine Software zu implementieren?

Sauttets committed:
Merge branch 'evcc-io:master' into master

Sauttets committed:
Merge branch 'evcc-io:master' into pulsatrix-Charger

Sauttets committed:
Merge pull request #4 from Sauttets/pulsatrix-Charger

Pulsatrix charger

andig commented:
Probiers bitte nochmal, PR ist wieder offen. Dann würde ich einmal drüber gehen und Grundsätzliches korrigieren.

Sauttets commented:
Der Haken bei "Allow edits by maintainers" ist gesetzt eigentlich müsstest du Sachen ändern können

andig committed:
Cleanup

andig committed:
no message

Sauttets committed:
Update Pulsatrix charger data structure and fix
websocket connection bug

Sauttets committed:
Merge pull request #5 from Sauttets/px-Charger

Update Pulsatrix charger data structure and fix

Sauttets commented:
Ich habe mir den Code jetzt angeschaut, die valid Funktion komplett entfernen können und ausgiebig getestet und er scheint das zu machen was er soll. Hattest du etwas spezielles im Sinn mit "Die read/write Logik braucht noch mehr Arbeit"? Die funktioniert aktuell so wie soll

Sauttets committed:
Merge branch 'master' into master

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,258 @@
+// Heartbeat sends a heartbeat to the pulsatrix SECC
+func (c *Pulsatrix) heartbeat() {
+	for range time.Tick(3 * time.Minute) {
+		if err := c.Enable(c.enabled); err != nil {
```
Muss hier geschrieben werden oder reicht lesen?

Sauttets replied:
Lesen reicht leider nicht aus

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,258 @@
+		messageType, message, err := c.conn.Read(ctx)
+		if err != nil {
+			c.log.ERROR.Println("error reading message:", err)
+			break
```
Was passiert, wenn der Loop mal verlassen wird?

Sauttets replied:
Dann sollte nach meiner Logik die Ws-Verbindung geschlossen werden, die go routine beendet und nach dem reconnect wieder neu gestartet.

andig replied:
Wo passiert der Neustart?

Sauttets replied:
reconnect ruft die connect funktion auf, die die go routine starte (Zeile 90). Allerdings fällt mir gerade auf das der Heartbeat auch nochmal gestartet wird obwohl der alte noch läuft. Da sollte ich auf jeden fall nochmal was dran machen

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,258 @@
+		}
+	}
+
+	c.mu.Lock()
```
Wie könnte das hier nil sein?

Sauttets replied:
Garnicht, Altlast

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,258 @@
+			return err
+		}
+	} else {
+		return fmt.Errorf("Waiting for reconnect")
```
Das wird im reconnect schon geloggt- hier raus.

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,258 @@
+// wsWrite writes a message to the websocket
+func (c *Pulsatrix) write(message string) error {
+	if c.conn != nil {
+		c.mu.Lock()
```
Wozu? Wenn braucht es das oben um auf die asynchron geänderte conn zuzugreifen.

Sauttets commented:
Was sollte aus deiner Sicht das Verhalten sein, wenn ein Reconnect, aus welchem Grund auch immer, über längere Zeit nicht funktioniert? Die Get-Funktionen liefern dann konstant Fehlermeldungen, werden aber trotzdem weiter aufgerufen.

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
minor fixes

andig commented:
> Was sollte aus deiner Sicht das Verhalten sein, wenn ein Reconnect, aus welchem Grund auch immer, über längere Zeit nicht funktioniert? Die Get-Funktionen liefern dann konstant Fehlermeldungen, werden aber trotzdem weiter aufgerufen.

Das ist dann halt so- Verbindung ist ja nicht da.

Sauttets committed:
Merge branch 'evcc-io:master' into master

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,254 @@
+	if err := backoff.RetryNotify(c.connectWs, bo, func(err error, time time.Duration) {
+		c.log.WARN.Printf("trying to reconnect in %v...\n", time)
+	}); err != nil {
+		c.log.ERROR.Println("RetryNotify: ", err)
```
Brauchts das? Man sieht ja beim Datenabruf dass keine Verbindung da ist.

Sauttets replied:
Nein, das wird sowieso nicht gelogged, da das nur passieren würde wenn er nicht weiter RetryNotify aufruft, das macht er aber unendlich lange. 
Ist raus geflogen

andig replied:
Warum nicht `backoff.Retry`?

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,254 @@
+	bo.MaxInterval = 5 * time.Minute
+
+	if err := backoff.RetryNotify(c.connectWs, bo, func(err error, time time.Duration) {
+		c.log.WARN.Printf("trying to reconnect in %v...\n", time)
```
Das wäre DEBUG oder sogar eher TRACE

andig commented on charger/pulsatrix.go:
```diff
@@ -0,0 +1,254 @@
+// ReconnectWs reconnects to a pulsatrix SECC websocket
+func (c *Pulsatrix) reconnectWs() {
+	bo := backoff.NewExponentialBackOff()
+	bo.InitialInterval = 30 * time.Second
```
Ich würde beide kürzer machen. 1s Start und 1m längstens. Damit kann man keinen Server lahm legen.

Sauttets committed:
Update pulsatrix.go

Changed heartbeat logic to close after connection loss using channels; removed unnecessary data; changed timing

Sauttets committed:
Update pulsatrix.go

error handling for lint

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
backoff RetryNotify -> Retry

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets commented:
Der code läuft aktuell stabil, allerdings besteht noch ein kleines Problem, für das ich noch keine schöne Lösung finden konnte:

Wenn man einen pulsatrix Ladepunkt mit dem wizard konfiguriert, bekommt man bei einem funktionierenden Ladepunkt folgenden error:

```
Teste die Konfiguration von pulsatrix ...
   Fehler: outdated
```
Der Fehler entsteht, da die WebSocket Nachricht zum Zeitpunkt des Status Funktionsaufrufes noch nicht durch den parser durch ist.
Die einzige Lösung die ich finden konnte initialisiert den VehicleStatus wert direkt nach dem eine WebSocket-Verbindung aufgebaut wurde:
```go
c.parseWsMessage(1, []byte("{\"vehicleStatus\": \"A1\"}"))
```
aber das würde ich ungerne ich so nicht implementieren, weil hässlich.

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
updated parser

due to a changed WebSocket message, the parser needed some adjustments

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
added support for 1p3p

Sauttets committed:
changed ws endpoint name

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
Merge branch 'master' into master

Sauttets committed:
added MIT license

andig committed:
Cleanup

andig committed:
wip

andig committed:
Simplify
